### PR TITLE
Fix docs

### PR DESCRIPTION
**Pull request by SergeyTsimfer**
This PR:

- Adds and improves a lot of docstrings in `TFModel` and `TorchModel`

- Fixes warnings during docs autogeneration by improving docs of `ImagesBatch`, `Pipeline`, `Research` and `NamedExpressions`. See #414 

- Changes the style of doc to `sphinx_rtd_theme`

### CODE REVIEW

File: batchflow/models/torch/base.py [hunk 1]
```diff
@@ -143,60 +165,58 @@ class TorchModel(BaseModel):
...
 
-    Things worth mentioning:
 
-    #. Input data and its parameters should be defined in configuration under ``inputs`` key.
-       See :meth:`.TorchModel._make_inputs` for details.
+    **In order to use existing model, it is recommended to configure:**
 
-    #. You might want to use a convenient multidimensional :class:`.ConvBlock`,
-       as well as other predefined layers from ``dataset.models.torch.layers``.
-       Of course, you can use usual `Torch layers <https://pytorch.org/docs/stable/nn.html>`_.
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/base.py]: Hi

File: batchflow/models/tf/base.py [hunk 2]
```diff
@@ -1567,21 +1539,24 @@ def output(self, inputs, predictions=None, ops=None, prefix=None, **kwargs):
             input tensors
 
         predictions : str or callable
```
Comment (Dimonovez) [inline batchflow/models/tf/base.py]:
```suggestion
         predictions : {'sigmoid', 'proba', 'labels', 'softplus'} or callable
```

File: batchflow/models/tf/base.py [hunk 3]
```diff
@@ -1689,7 +1663,7 @@ def _add_output_callable(self, inputs, oper, name, attr_prefix, device, **kwargs
 
     @classmethod
     def default_config(cls):
-        """ Define model defaults
+        """ Define model defaultsю
```
Comment (Dimonovez) [inline batchflow/models/tf/base.py]:
```suggestion
        """ Define model defaults.
```

**Comment by Dimonovez** [comment]
Do we have clear rules for using one ` or two ``?

File: batchflow/models/tf/base.py [hunk 4]
```diff
@@ -97,27 +119,51 @@ class TFModel(BaseModel):
...
         - ``{'decay': 'exp'}``
         - ``{'decay': ('polynomial_decay', {'decay_steps': 10000})}``
         - ``{'decay': {'name': tf.train.inverse_time_decay, 'decay_rate': .5}``
 
-    scope - subset of variables to optimize during training. Can be either string or sequence of strings.
-        Value `''` is reserved for optimizing all trainable variables.
-        Putting `-` sign before name stands for complement: optimize everything but the passed scope.
+    scope : str or sequence of str
+        Subset of variables to optimize during training.
+        Value ``''`` is reserved for optimizing all trainable variables.
```
Comment (Dimonovez) [inline batchflow/models/tf/base.py]: Do we have clear rules for using single and double back-ticks?

File: batchflow/named_expr.py [hunk 5]
```diff
@@ -656,16 +656,17 @@ class I(NamedExpression):
     Parameters
     ----------
     name : str
```
Comment (Dimonovez) [inline batchflow/named_expr.py]: Are substrings needed? Can `name` type be a fixed set of values?

File: batchflow/models/tf/base.py (hunk 4 above)
Comment (SergeyTsimfer) [reply batchflow/models/tf/base.py]: We do!
Double backquotes are reserved for code samples (or things that can appear generally in config), while single ones are used for emphasis

File: batchflow/named_expr.py (hunk 5 above)
Comment (SergeyTsimfer) [reply batchflow/named_expr.py]: This set is too big: `current`/`maximum`/`ratio` or any of the substrings

File: batchflow/models/tf/base.py [hunk 6]
```diff
@@ -1472,18 +1439,19 @@ def _dynamic_crop(cls, inputs, static_shape, dynamic_shape, data_format='channel
 
     @classmethod
     def initial_block(cls, inputs, name='initial_block', **kwargs):
-        """ Transform inputs with a convolution block
+        """ Transform inputs with a convolution block. Usually used for initial preprocessing,
```
Comment (a-arefina) [inline batchflow/models/tf/base.py]: IMO convolution is a detail of realisation here
```suggestion
        """ Transform inputs. Usually used for initial preprocessing,
```

File: batchflow/models/tf/base.py [hunk 7]
```diff
@@ -1643,15 +1618,15 @@ def output(self, inputs, predictions=None, ops=None, prefix=None, **kwargs):
         However, if one of the placeholders also has a name 'labels', then it will be lost as the model
         will rewrite the name 'labels' with an output.
 
-        That is where a dict might be convenient::
+        That is where a dict might be convenient:
```
Comment (a-arefina) [inline batchflow/models/tf/base.py]:
```suggestion
        In this case dict might be more convenient:
```
+merge this with previous paragraph

File: batchflow/models/tf/base.py [hunk 8]
```diff
@@ -1777,26 +1754,21 @@ def fill_params(cls, _name, **kwargs):
...
 
     def build_config(self, config=None):
-        """ Define a model architecture configuration
-
-        It takes just 2 steps:
-
-        #. Define names for all placeholders and make input tensors by calling ``super().build_config(names)``.
-
-           If the model config does not contain any name from ``names``, :exc:`KeyError` is raised.
+        """ Define a model architecture configuration.
```
Comment (a-arefina) [inline batchflow/models/tf/base.py]:
```suggestion
        """ Define model's architecture configuration.
```

File: batchflow/models/tf/base.py [hunk 9]
```diff
@@ -1777,26 +1754,21 @@ def fill_params(cls, _name, **kwargs):
...
-           See :meth:`.TFModel._make_inputs` for details.
+        * Don't forget to call ``super().build_config(names)`` in the beginning.
 
-        #. Define parameters for :meth:`.TFModel.initial_block`, :meth:`.TFModel.body`, :meth:`.TFModel.head`
-           which depend on inputs.
+        * Define parameters for :meth:`.TFModel.initial_block`, :meth:`.TFModel.body`, :meth:`.TFModel.head`,
+          which depend on inputs.
 
-        #. Don't forget to return ``config``.
+        * Dont forget to return ``config`` at the end.
```
Comment (a-arefina) [inline batchflow/models/tf/base.py]:
```suggestion
        * Don't forget to return ``config`` at the end.
```

File: batchflow/models/tf/base.py [hunk 10]
```diff
@@ -1878,28 +1850,28 @@ def data_format(self, tensor, **kwargs):
...
 
     def num_classes(self, tensor):
-        """ Return the  number of classes """
+        """ Return the number of classes. """
         if self.has_classes(tensor):
             classes = self.classes(tensor)
             return classes if isinstance(classes, int) else len(classes)
         return self.get_num_channels(tensor)
 
     def get_num_channels(self, tensor, **kwargs):
```
Comment (a-arefina) [inline batchflow/models/tf/base.py]: why `num_classes` but `get_num_channels`?...

File: batchflow/models/tf/base.py [hunk 11]
```diff
@@ -1916,7 +1888,7 @@ def get_num_channels(self, tensor, **kwargs):
 
     @classmethod
     def num_channels(cls, tensor, data_format='channels_last'):
-        """ Return number of channels in the input tensor
+        """ Return number of channels in the input tensor.
```
Comment (a-arefina) [inline batchflow/models/tf/base.py]: ok, what's the difference between `get_num_channels` and `num_channels`? Can we leave only one method?

File: batchflow/models/tf/base.py [hunk 12]
```diff
@@ -2148,19 +2112,18 @@ def scse_block(cls, inputs, ratio=2, name='scse', **kwargs):
...
         ----------
         inputs : tf.Tensor or tuple of two tf.Tensor
-            a tensor to resize and a tensor which size to resize to
+            Tensor to resize.
         factor : int
-            an upsamping scale
+            Upsamping scale.
         resize_to : tf.Tensor
-            a tensor which shape the output should be resized to
+            Tensor which shape the output should be resized to.
```
Comment (a-arefina) [inline batchflow/models/tf/base.py]:
```suggestion
            Tensor which shape is used to resize the output.
```

File: batchflow/models/torch/base.py [hunk 13]
```diff
@@ -49,28 +49,48 @@
 
 
 class TorchModel(BaseModel):
-    r""" Base class for torch models
+    r""" Base class for Torch models.
```
Comment (a-arefina) [inline batchflow/models/torch/base.py]: is `r` prefix needed?

File: batchflow/models/torch/base.py [hunk 14]
```diff
@@ -49,28 +49,48 @@
...
 
-    **Configuration**
+    Parameters
+    ----------
+    inputs : dict
+        Mapping from placeholder names (e.g. ``images``, ``labels``, ``masks``) to arguments of its initialization.
```
Comment (a-arefina) [inline batchflow/models/torch/base.py]: See TFModel's comments

File: batchflow/models/torch/base.py [hunk 15]
```diff
@@ -459,21 +480,23 @@ def num_classes(self, tensor):
 
     @classmethod
     def default_config(cls):
-        """ Define model defaults
+        """ Define model defaultsю
```
Comment (a-arefina) [inline batchflow/models/torch/base.py]:
```suggestion
        """ Define model defaults.
```

File: batchflow/models/torch/base.py [hunk 16]
```diff
@@ -459,21 +480,23 @@ def num_classes(self, tensor):
...
 
         You need to override this method if you expect your model or its blocks to serve as a base for other models
         (e.g. VGG for FCN, ResNet for LinkNet, etc).
 
         Put here all constants (like the number of filters, kernel sizes, block layouts, strides, etc)
         specific to the model, but independent of anything else (like image shapes, number of classes, etc).
 
-        These defaults can be changed in :meth:`~.TorchModel.build_config` or when calling :meth:`.Pipeline.init_model`.
+        These defaults can be changed in :meth:`~.TFModel.build_config` or when calling :meth:`.Pipeline.init_model`.
```
Comment (a-arefina) [inline batchflow/models/torch/base.py]: undo

File: batchflow/models/torch/base.py [hunk 17]
```diff
@@ -459,21 +480,23 @@ def num_classes(self, tensor):
...
 
-        Usually, it looks like::
+        Examples
+        --------
+        .. code-block:: python
 
             @classmethod
             def default_config(cls):
-                config = TorchModel.default_config()
+                config = TFModel.default_config()
```
Comment (a-arefina) [inline batchflow/models/torch/base.py]: undo

File: batchflow/models/torch/base.py [hunk 18]
```diff
@@ -503,27 +526,22 @@ def get_defaults(cls, name, kwargs):
...
-        It takes just 2 steps:
-
-        #. Define names for input data and make input tensors by calling ``super().build_config(names)``.
+        """ Define a model architecture configuration.
 
-           If the model config does not contain any name from ``names``, :exc:`KeyError` is raised.
+        * Don't forget to call ``super().build_config(names)`` in the beginning.
 
-           See :meth:`._TorchModel.make_inputs` for details.
+        * Define parameters for :meth:`.TFModel.initial_block`, :meth:`.TFModel.body`, :meth:`.TFModel.head`,
```
Comment (a-arefina) [inline batchflow/models/torch/base.py]:
```suggestion
        * Define parameters for :meth:`.Torch.initial_block`, :meth:`.Torch.body`, :meth:`.Torch.head`,
```

File: batchflow/models/torch/base.py [hunk 19]
```diff
@@ -503,27 +526,22 @@ def get_defaults(cls, name, kwargs):
...
-        Typically it looks like this::
+        Examples
+        --------
+        .. code-block:: python
 
             def build_config(self, names=None):
-                names = names or ['images', 'labels']
                 config = super().build_config(names)
-                config['head/num_classes'] = self.num_classes('targets')
+                config['head']['num_classes'] = self.num_classes('targets')
```
Comment (a-arefina) [inline batchflow/models/torch/base.py]: I like `config['head/num_classes']` style

File: batchflow/models/torch/base.py [hunk 20]
```diff
@@ -564,11 +582,12 @@ def _build(self, config=None):
 
     @classmethod
     def initial_block(cls, **kwargs):
-        """ Transform inputs with a convolution block
+        """ Transform inputs with a convolution block. Usually used for initial preprocessing,
```
Comment (a-arefina) [inline batchflow/models/torch/base.py]:
```suggestion
        """ Transform inputs. Usually used for initial preprocessing,
```

File: batchflow/models/torch/base.py [hunk 21]
```diff
@@ -655,15 +677,15 @@ def output(self, inputs, predictions=None, ops=None, prefix=None, **kwargs):
         However, if one of the placeholders also has a name 'labels', then it will be lost as the model
         will rewrite the name 'labels' with an output.
 
-        That is where a dict might be convenient::
+        That is where a dict might be convenient:
```
Comment (a-arefina) [inline batchflow/models/torch/base.py]: See corresponding TF comment

File: batchflow/models/torch/base.py [hunk 22]
```diff
@@ -780,27 +802,25 @@ def train(self, *args, fetches=None, use_lock=False, microbatch=None):    # pyli
...
+            Sequence of tensor names to calculate and return.
 
         use_lock : bool
-            if True, the whole train step is locked, thus allowing for multithreading.
+            If True, the whole train step is locked, thus allowing for multithreading.
 
         microbatch : int or None
-            make forward/backward pass with microbatches of a given size, but apply gradients after the whole batch.
-            Batch size should be evenly divisible by microbatch size.
+            Size of chunks to split every batch in. Allows to process given data sequentially, accumulating gradients
```
Comment (a-arefina) [inline batchflow/models/torch/base.py]:
```suggestion
            Size of chunks to split every batch into. Allows to process given data sequentially, accumulating gradients
```

File: batchflow/models/torch/base.py [hunk 23]
```diff
@@ -780,27 +802,25 @@ def train(self, *args, fetches=None, use_lock=False, microbatch=None):    # pyli
...
+            from microbatches and applying them once in the end.
 
         Returns
         -------
-        Calculated values of tensors in `fetches` in the same structure
-
+        Calculated values of tensors in `fetches` in the same structure.
```
Comment (a-arefina) [inline batchflow/models/torch/base.py]:
```suggestion
        Calculated values of tensors in `fetches` in the same order.
```

File: batchflow/models/torch/base.py [hunk 24]
```diff
@@ -859,35 +879,29 @@ def train(self, *args, fetches=None, use_lock=False, microbatch=None):    # pyli
...
-        microbatch : int or None
-            make forward/backward pass with microbatches of a given size, but apply gradients after the whole batch.
-            Batch size should be evenly divisible by microbatch size.
+            If True, the whole train step is locked, thus allowing for multithreading.
 
         Returns
         -------
-        Calculated values of tensors in `fetches` in the same structure
-
+        Calculated values of tensors in `fetches` in the same structure.
```
Comment (a-arefina) [inline batchflow/models/torch/base.py]:
```suggestion
        Calculated values of tensors in `fetches` in the same order.
```

File: batchflow/research/research.py [hunk 25]
```diff
@@ -144,12 +144,11 @@ def add_function(self, function, returns=None, name=None, execute=1, dump='last'
         Parameters
         ----------
         function : callable
-            callable object with following parameters:
+            callable object. must include with following named parameters:
```
Comment (a-arefina) [inline batchflow/research/research.py]:
```suggestion
            callable object. must include following named parameters:
```

File: batchflow/research/research.py [hunk 26]
```diff
@@ -468,16 +467,29 @@ class Executable:
         or returns (for function) values are lists of variable values
     path : str
         path to the folder where results will be dumped
-    exec : int, list of ints or None
+    execute : int, list of ints or None
```
Comment (a-arefina) [inline batchflow/research/research.py]: Does all this include @AlexeyKozhevin's research PR?

File: batchflow/research/research.py [hunk 27]
```diff
@@ -468,16 +467,29 @@ class Executable:
...
+        If `'last'`, function will be executed just at last iteration (if `iteration + 1 == n_iters`
+        or `StopIteration` was raised)
+
+        If positive int, function will be executed each `step` iterations.
+
+        If str, must be `'#{it}'` or `'last'` where it is int,
```
Comment (a-arefina) [inline batchflow/research/research.py]:
```suggestion
        If str, must be `'#{it}'` or `'last'` where `{it}` is int,
```

File: docs/conf.py [hunk 28]
```diff
@@ -76,43 +76,40 @@
...
-html_theme = 'bizstyle'
-#html_theme = 'sphinx_rtd_theme'
+# choose html theme with custom css template
+### haiku, classic, bizstyle, sphinx_rtd_theme
+html_theme = 'sphinx_rtd_theme'
+
 
 html_theme_options = {
-#    "rightsidebar": "false"
+   # "rightsidebar": "false"
```
Comment (a-arefina) [inline docs/conf.py]: not aligned now

File: docs/conf.py [hunk 29]
```diff
@@ -76,43 +76,40 @@
...
+#     "page_width": 'auto',
+#     # "sidebar_width": '20vw'
+# }
+# html_sidebars = {
+#     '**': [
+#         'about.html',
+#         'navigation.html',
+#         'searchbox.html',
+#     ]
+# }
```
Comment (a-arefina) [inline docs/conf.py]: Do we need this in repo?

File: docs/intro/torch_models.rst [hunk 30]
```diff
@@ -239,7 +239,7 @@ For more detail see :class:`.TorchModel` documentation.
 How to write a custom model
 ===========================
 
-To begin with, take a look at :class:`.ConvBlock` to find out how to write complex networks in just one line of code.
+To begin with, take a look at :class:`~.torch.layers.ConvBlock` to find out how to write complex networks in just one line of code.
```
Comment (a-arefina) [inline docs/intro/torch_models.rst]: do we need similar changes in TF documentation?

File: batchflow/models/tf/base.py [hunk 31]
```diff
@@ -44,48 +44,71 @@
...
-        model inputs (see :meth:`.TFModel._make_inputs`)
-
-    loss - a loss function, might be defined in one of three formats:
-        - name
-        - tuple (name, args)
-        - dict {'name': name, \**args}
-
-        where name might be one of:
-            - short name (`'mse'`, `'ce'`, `'l1'`, `'cos'`, `'hinge'`, `'huber'`, `'logloss'`, `'dice'`)
+        Mapping from placeholder names (e.g. ``images``, ``labels``, ``masks``) to arguments of its initialization.
```
[the remark this answers is not in the document]
Comment (SergeyTsimfer) [reply batchflow/models/tf/base.py]: Placeholders!

File: batchflow/models/tf/base.py [hunk 32]
```diff
@@ -44,48 +44,71 @@
...
+        Allows to create placeholders of needed format (shape, dtype, data format) with specified name
+        and apply some typical transformations (like one-hot-encoding), if needed.
+
+        If value is a string, then it must point to another key from the input-dict, effectively making an alias.
+        By default, ``targets`` is aliased to ``labels`` or ``masks``, if present.
+        If value is a tuple, then it must contain all of the arguments below in the same order. ``None`` is reserved
+        for using default value.
+        If value is a dictionary, then it can omit some of the parameters (default values will be at use).
```
[the remark this answers is not in the document]
Comment (SergeyTsimfer) [reply batchflow/models/tf/base.py]: It is not possible type of `input`, it is possible type of `value` inside the `inputs`-dictionary. Added clarity

File: batchflow/models/tf/base.py [hunk 33]
```diff
@@ -126,116 +173,118 @@ class TFModel(BaseModel):
...
-       by :meth:`~.TFModel.get_from_attr`.
 
-    #. For decay and training control there is a predefined ``global_step`` tensor. You can get it
-       by :meth:`~.TFModel.get_from_attr`.
+    **In order to use existing model, it is recommended to define following keys in configuration dictionary:**
 
-    #. In many cases there is no need to write a loss function, learning rate decay and optimizer
-       as they might be defined through config.
+    * ``inputs``: defines input data together with parameters like shape, dtype, number of classes.
+      See :meth:`.TFModel._make_inputs` for details.
```
[the remark this answers is not in the document]
Comment (SergeyTsimfer) [reply batchflow/models/tf/base.py]: That's a shame

File: batchflow/models/tf/base.py [hunk 34]
```diff
@@ -1042,39 +1014,41 @@ def train(self, fetches=None, feed_dict=None, use_lock=True, train_mode='', micr
...
 
         Returns
         -------
-        Calculated values of tensors in `fetches` in the same structure
-
-        See also
-        --------
-        `Tensorflow Session run <https://www.tensorflow.org/api_docs/python/tf/Session#run>`_
+        tuple, list
+            Calculated values of tensors in `fetches` in the same structure.
```
[the remark this answers is not in the document]
Comment (SergeyTsimfer) [reply batchflow/models/tf/base.py]: It is structure: see [Session.run](https://www.tensorflow.org/versions/r1.14/api_docs/python/tf/Session#run)

File: batchflow/models/tf/base.py (hunk 11 above)
Comment (SergeyTsimfer) [reply batchflow/models/tf/base.py]: All of the `get_###` methods rely on `tensor_config`, which is created and stored for some tensors during `_make_inputs`. That allows to better control complex networks with multiple inputs of different `data_formats`, for example

File: batchflow/research/research.py (hunk 26 above)
Comment (SergeyTsimfer) [reply batchflow/research/research.py]: It does not; all of the fixes are to get rid of `sphinx` warnings

File: batchflow/batch_image.py [hunk 35]
```diff
@@ -475,22 +492,26 @@ def _scale_(self, image, factor, preserve_shape=False, origin='center', resample
             Relevant only if `preserve_shape` is True.
             If `scale` < 1, defines position of the scaled image with respect to the original one's shape.
             If `scale` > 1, defines position of cropping box.
+
+            Can be one of:
```
[the remark this answers is not in the document]
Comment (SergeyTsimfer) [reply batchflow/batch_image.py]: Done

**Pull request merged**